### DISCUSSION TITLE
Revisiting Unions and using a source generator to get us 70% of the way to a solution

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#5244] [Language Ideas]
(Tagging @333fred here as he previously showed interest in this topic and some of these ideas may be of use to him)

Some time ago, I started the discussion [Discriminated Unions features "shopping list"](https://github.com/dotnet/csharplang/discussions/3760), in which I proposed various ideas around how C# might implement discriminated unions (DUs). A year later, I've been experimenting with some ideas and I now have a solution to what was the biggest challenge for me at the time: how to have a closed hierarchy of types, especially when using structs.

The code examples I'll use below come from https://github.com/DavidArno/CSharpUnions, which I'm using to currently experiment with the shape of the generated union types as I'm planning on providing a source-generator based solution to creating these DUs.

Let's start with one of the example DUs that I talked about last year:
```
enum struct Shapes 
{
    Point,  
    Rectangle { double width, double length },
    Circle { double radius }
}
```

As I'm working on a source generator that can create such a DU, we need a definition for the generator to work with:
```c#
[DiscriminatedUnion]
record ShapeDefinition
{
    record Point();

    record Rectangle(double Width, double Length);

    record Circle(double Radius);
}
```

And - at the time of writing this - I'm anticipating the generator would create something like:
```c#
public readonly record struct Point();

public readonly record struct Rectangle(double Width, double Length);

public readonly record struct Circle(double Radius);

public readonly struct Shape
{

    private enum Cases { PointCase, RectangleCase, CircleCase }

    private readonly Point _casePoint;
    private readonly Rectangle _caseRectangle;
    private readonly Circle _caseCircle;

    private readonly Cases _validCase;

    private readonly Type<Point> _typePoint;
    private readonly Type<Rectangle> _typeRectangle;
    private readonly Type<Circle> _typeCircle;

    private readonly bool _initialised;

    private Shape(Point casePoint, Rectangle caseRectangle, Circle caseCircle, Cases validCase)
    {
        _casePoint = casePoint;
        _caseRectangle = caseRectangle;
        _caseCircle = caseCircle;

        _validCase = validCase;

        _typePoint = Type<Point>.Instance;
        _typeRectangle = Type<Rectangle>.Instance;
        _typeCircle = Type<Circle>.Instance;

        _initialised = true;
    }

    public static Shape AsPoint() => new(new Point(), default, default, Cases.PointCase);

    public static Shape AsRectangle(double width, double length) =>
        new(default, new Rectangle(width, length), default, Cases.RectangleCase);

    public static Shape AsCircle(double radius) => new(default, default, new Circle(radius), Cases.CircleCase);

    public object Case => _validCase switch {
        Cases.PointCase => _typePoint,
        Cases.RectangleCase => _typeRectangle,
        Cases.CircleCase or _ => _typeCircle
    };

    public Point Point => _casePoint;

    public Rectangle Rectangle => _caseRectangle;

    public Circle Circle => _caseCircle;


    public void Deconstruct(out object type, out object? value)
    {
        if (!_initialised) throw null!;

        switch (_validCase)
        {
            case Cases.PointCase:
                type = _typePoint;
                value = _casePoint;
                break;
            case Cases.RectangleCase:
                type = _typeRectangle;
                value = (_caseRectangle.Width, _caseRectangle.Length);
                break;
            case Cases.CircleCase:
            default:
                type = _typeCircle;
                value = _caseCircle.Radius;
                break;
        }
    }
}
```
By why of demonstrating how the above might be used, here's a test of one of the ways of pattern matching this DU:
```c#
[Test]
public void ForRectangeCase_CanPatternMatchWidthAndLength()
{
    var shape = Shape.AsRectangle(10, 20);

    var area = shape switch {
        { Case: Type<Point> } => 0.0,
        { Case: Type<Rectangle>, Rectangle: { Width: var w, Length: var l } } => w * l,
        { Case: Type<Circle>, Circle.Radius: var r } => Math.PI * r * r
    };

    Assert.AreEqual(200, area);
}
```
The "breakthrough" I made that had me stuck for so long was realising that there was no reason to view `Rectangle` as a subtype of `Shape`. A `Shape` DU may contain a `Rectangle` value (`Case` as I've referred to it), but "contains" is the key here: it isn't a Rectangle; it contains one. 

The moment we move from `Rectangle` being an aspect of `Shape` itself to it just being data that `Shape` contains, then the problem of how can a type also be another type without inheritance goes away. As a result, we can just use standard property patterns to pattern match the DU.

In the title, I say this covers 70% of the functionality of what we might expect from C# DUs. That's because I see three things missing from any solution I might create, which a compiler lowering-based solution could resolve.

## Issue 1:  `Type<Rectangle>` ##
`typeof(Rectangle)`  isn't a constant and so it cannot be used in a pattern. [The proposal switch on System.Type](https://github.com/dotnet/csharplang/discussions/356) has been around for a long time and this is another example of how supporting that would make things simpler when matching types. My solution is to use a custom `Type<T>` and for the `Case` property to return `object` so that I can match the type that way. But directly testing the type in a pattern would improve readability:
```c#
var area = shape switch {
    { Case: Point } => 0.0,
    { Case: Rectangle, Rectangle: { Width: var w, Length: var l } } => w * l,
    { Case: Circle, Circle.Radius: var r } => Math.PI * r * r
};
```
## Issue 2: Constraining the properties ##
There's nothing to stop me doing:
```cs
var rectArea = shape is { Case: Type<Rectangle>, Circle.Radius: var r } ? Math.PI * r * r : 0.0;
```
The compiler ought to be able to treat accessing properties not applicable to the case as an error.
## Issue 3: Match completeness checking ##
My test compiles without warning as a stuck a `#pragma warning disable CS8509` in the code. As `Case` is of type `object`, clearly the three cases I have aren't exhaustive of the value it could take - as far as the compiler knows. This of course relates back to0 the whole idea of viewing DUs as one form of closed enum. For:
```
enum struct Shapes 
{
    Point,  
    Rectangle { double width, double length },
    Circle { double radius }
}
```
the compiler can know that there are only three cases and could therefore determine that the switch expression is indeed exhaustive.

**Comment by 333fred** [comment]
Thanks for the ping David, your experiment was good reading. In particular, this comment:

> The "breakthrough" I made that had me stuck for so long was realising that there was no reason to view Rectangle as a subtype of Shape. A Shape DU may contain a Rectangle value (Case as I've referred to it), but "contains" is the key here: it isn't a Rectangle; it contains one.

That's an interesting way to approach the problem. One immediate concern I would have with that approach, however, is that since we're talking about a struct here, `Shape` becomes as big as the combination of all 3 sub-values. If we have some form of struct inheritance, then we could have them overlap in memory and reduce the footprint. Maybe you could do something with explicit layouts? It's not particularly safe if issue 2 isn't resolved though, as you could easily observe sheared values.

**Reply by HaloFour** [reply]
> since we're talking about a `struct` here, Shape becomes as big as the combination of all 3 sub-values

This is how struct DUs work in F#, has that proven to be a problem?

**Reply by 333fred** [reply]
> This is how struct DUs work in F#, has that proven to be a problem?

In general, I want to make sure that when we make new general-use language features, our core customers can use them. The BCL itself is one of those core customers, and I can pretty easily guarantee that they would avoid DU's if they ballooned struct sizes.

**Reply by HaloFour** [reply]
By "proven" I meant with numbers.  Don't get me wrong, I'm not suggesting that the compiler _shouldn't_ optimize where it can here, I only ask whether doing so is worth the return.  😁

**Reply by DavidArno** [reply]
@333fred,

I'm happy to experiment with overlapping. One thought though, am I correct in thinking this is really only needed for non-readonly structs? I'm thinking that the size of a readonly struct isn't so important as it's passed by reference. But I may have that wrong.

There again, if overlapping is used, it's probably easier to do it for all structs.

**Reply by 333fred** [reply]
> By "proven" I meant with numbers. Don't get me wrong, I'm not suggesting that the compiler shouldn't optimize where it can here, I only ask whether doing so is worth the return

Framework guidelines say anything bigger than 16 bytes should be a class or always passed by reference. This can change based on native platform slightly, but the general idea is that the copies become inefficient and register passing becomes impossible. In a language feature that we'd like to be a core concept of the language, building in inefficient representation should be priority to avoid.

> am I correct in thinking this is really only needed for non-readonly structs?

Rather than readonly structs, I would phrase it as structs that are intended to be passed by value. Result, for example, I would almost always expect to be passed by value, but we wouldn't want it to be bigger than the biggest of its operands (plus a discriminant).

**Reply by HaloFour** [reply]
@333fred 

> Framework guidelines say anything bigger than 16 bytes should be a class or always passed by reference.

That's fair.  I am curious, have conversations with the BCL team about DUs taken place?  Has there been interest in official DUs to be used by BCL APIs?  I'm sure there would be a lot of interest in whether the BCL would adopt DUs, especially those general purpose ones like `Maybe<T>` or `Result<T>`.

**Reply by 333fred** [reply]
> I am curious, have conversations with the BCL team about DUs taken place?

In depth discussions? Not to my knowledge. We've certainly talked a bit about doing so here and there.
